spinalcordtoolbox/csa_pmj.py

```diff
@@ -0,0 +1,206 @@
+    """
+    Compute distance from PMJ projection on centerline for all the centerline.
+    Generate mask from segmentation of the slices used to process segmentation data corresponding to a distance from PMJ projection.
+    :param segmentation: input segmentation. Could be either an Image or a file name.
+    :param pmj: label of PMJ.
+    :param distance: float: Distance from Ponto-Medullary Junction (PMJ) in mm.
+    :param param_centerline: see centerline.core.ParamCenterline()
+    :param verbose:
+    :return mask:
+    :return slices:
+
+    """
+    native_orientation = Image(segmentation).orientation
+    im_seg = Image(segmentation).change_orientation('RPI')
+    im_pmj = Image(pmj).change_orientation('RPI')
+    nx, ny, nz, nt, px, py, pz, pt = im_seg.dim
+    pr = min([px, py])
+    im_segr = resample_nib(im_seg, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    im_pmjr = resample_nib(im_pmj, new_size=[pr, pr, pz], new_size_type='mm', interpolation='linear')
+    data_pmj = im_pmjr.data
+    # Update dimensions from resampled image
+    nx, ny, nz, nt, px, py, pz, pt = im_segr.dim
+
+    # Extract min and max index in Z direction
+    data_seg = im_segr.data
+    X, Y, Z = (data_seg > NEAR_ZERO_THRESHOLD).nonzero()
+    min_z_index = min(Z)
+    # Compute the spinal cord centerline based on the spinal cord segmentation
+    im_ctl, arr_ctl, arr_ctl_der, fit_results = get_centerline(im_segr, param=param_centerline, verbose=verbose)
+
+    # Extrapolate centerline
+    arr_ctl_extra = extrapolate_centerline(arr_ctl, im_segr, pz)
+
+    # Create an image with the centerline | ONLY TO VALIDATE --> to remove
+    im_centerline = im_segr.copy()
+    im_centerline.data = np.zeros(im_centerline.data.shape)
+    # Assign value=1 to centerline. Make sure to clip to avoid array overflow.
+    im_centerline.data[round_and_clip(arr_ctl_extra[0], clip=[0, im_centerline.data.shape[0]]),
+                       round_and_clip(arr_ctl_extra[1], clip=[0, im_centerline.data.shape[1]]),
+                       np.array(range(im_segr.dim[2]))] = 1
+    im_centerline.save('centerline_extrapolated.nii.gz')
+    pmj_coord = np.argwhere(data_pmj != 0)[0]
+    # Get Z index of PMJ project on extrapolated centerline
+    pmj_index = get_min_distance(pmj_coord, arr_ctl_extra, px, py, pz)
+    # Compute distance from PMJ along centerline
+    arr_length = get_distance_from_pmj(arr_ctl_extra, pmj_index, px, py, pz)
+
+    # Check if distance is out of bound
+    if distance > arr_length[0][0]:
+        raise ValueError("Input distance of " + str(distance) + " mm is out of bound for maximum distance of " + str(arr_length[0][0]) + " mm")
+
+    if distance < arr_length[0][-1]:
+        raise ValueError("Input distance of " + str(distance) + " mm is out of bound for minimum distance of " + str(arr_length[0][-1]) + " mm")
+
+    # Get Z index of corresponding distance from PMJ with the specified extent
```

Comment:
I ran the code in the debugger and found that the requested `extent=20` yielded the slices '183:207', which gives an extent of 24 (instead of 20). I haven't dug to much into the code but maybe there is an indexing issue somewhere?

Reply:
ahh, right now the extent is in mm (along the centerline) not in number of slices, what would be better?

Reply:
ahh i see, my mistake. We should definitely keep the extent in mm to be consistent with the distance from PMJ (and also to be agnostic to image resolution)